src/actions/app.js

```diff
@@ -416,3 +418,22 @@ export function updateBrowserConnectionStatus(
 ): Action {
   return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
 }
+
+export function toggleOpenCategoryInSidebar(
+  category: IndexIntoCategoryList
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const sidebarOpenCategories = new Set(getSidebarOpenCategories(getState()));
+    if (sidebarOpenCategories.has(category)) {
+      sidebarOpenCategories.delete(category);
+    } else {
+      sidebarOpenCategories.add(category);
+    }
+    withHistoryReplaceStateSync(() => {
+      dispatch({
+        type: 'CHANGE_OPEN_CATEGORIES',
+        sidebarOpenCategories: sidebarOpenCategories,
+      });
+    });
+  };
```

Comment:
looking at this now, I think this should move to the reducer.

Let me explain a bit:
* The reducers role is to change the state.
* This has an important limitation: it can't access other parts of the state. That's where using thunk actions like here can be useful. But in our case this is not needed.
* changing the state in reducer makes it easier to understand where / when / how / why some specific part of the state is changed.

So the action could look like this:
```js
export function toggleOpenCategoryInSidebar(
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    category,
  };
}
```

Then in the reducer:
```js
const sidebarOpenCategories: Reducer<Set<number>> = (
  state: Set<number> = new Set(),
  action
) => {
  switch (action.type) {
    case 'TOGGLE_SIDEBAR_OPEN_CATEGORY': {
      const newState = new Set(state);
      if (newState.has(category)) {
        newState.delete(category);
      } else {
        newState.add(category);
      }
      return newState;
    }
    default:
      return state;
  }
};
```

Finally the use of `withHistoryReplaceStateSync` would move up to the caller, in `_toggleCategory` in `CategoryBreakdown`.

Reply:
Sorry for the back and forth! I hope it makes sense

Reply:
It makes sense enough (I still have to read some more on this topic), thanks for the good explanation.